src/MessageScheduler.hpp

```diff
@@ -81,47 +84,60 @@ class MessageScheduler : public QObject {
    * Then, messages are sent by topic priority.
    */
   void schedule() {
-    if (!is_network_unblocked) {
+    if (network_backpressure_counter > max_queue_before_waiting) {
       return;
     }
 
+
     // flush no-drop queue
     if (!no_drop_queue.empty()) {
       while (!no_drop_queue.empty()) {
         const auto& next = no_drop_queue.front();
         Q_EMIT scheduled(next);
         no_drop_queue.pop_front();
+        network_backpressure_counter++;
+      }
+      if (network_backpressure_counter > max_queue_before_waiting) {
+        return;
       }
-      is_network_unblocked = false;
-      return;
     }
 
     if (topic_queue.empty()) {
       return;
     }
 
-    // update wait times for waiting messages.
-    // select topic with the highest wait time.
+    int messages_to_send = std::min((max_queue_before_waiting - network_backpressure_counter) + 1, topic_queue.size());
     const auto now = SchedulerClock::now();
-    auto next = topic_queue.begin();
+
+    // Collect candidates
+    std::vector<WaitingMessage*> candidates;
     for (auto it = topic_queue.begin(); it != topic_queue.end(); ++it) {
       const QString& candidate_topic = it->first;
       WaitingMessage& candidate = it->second;
 
       if (candidate.message_ready) {
-        candidate.time_waiting =
-            (now - candidate.last_send_time) * candidate.priority;
-      }
-      if (candidate.time_waiting > next->second.time_waiting) {
-        next = it;
+        candidate.time_waiting = (now - candidate.last_send_time) * candidate.priority;
       }
+      
+      candidates.push_back(&candidate);
```

Comment:
if you only push candidates that are ready to send, then you sort a smaller set of things and don't have to check message_ready again, right?